File: 78-subsets/subsets.cpp
class Solution {
public:
    void generateSubsets(int i,int n,vector<int>&subs,vector<int>&nums,vector<vector<int>>&ans)
    {
        if(i==n)
        {
            ans.push_back(subs);
            return;
        }

        subs.push_back(nums[i]);
        generateSubsets(i+1,n,subs,nums,ans);
        subs.pop_back();
        generateSubsets(i+1,n,subs,nums,ans);

    }
    vector<vector<int>> subsets(vector<int>& nums) 
    {
        vector<vector<int>>ans;
        vector<int>subs;

        generateSubsets(0,nums.size(),subs,nums,ans);
        return ans;
    }
};

File: 78-subsets/subsets.java
class Solution {
    public void findTotalSubsets(int index,int[]  nums,int n,List<List<Integer>>ans,List<Integer>subset)
    {
        if(index==n)
        {
            ans.add(new ArrayList<>(subset));
            return;
        }
        subset.add(nums[index]);
        findTotalSubsets(index+1,nums,n,ans,subset);
        subset.remove(subset.size()-1);
        findTotalSubsets(index+1,nums,n,ans,subset);
        
    }
    public List<List<Integer>> subsets(int[] nums) 
    {
        List<List<Integer>> ans=new ArrayList<>();
        List<Integer> subset= new ArrayList<>();

        findTotalSubsets(0,nums,nums.length,ans,subset);
        return ans;
    }
}